source/common/common/posix/thread_impl.cc

```diff
@@ -24,26 +26,89 @@ int64_t getCurrentThreadId() {
 
 } // namespace
 
-ThreadImplPosix::ThreadImplPosix(std::function<void()> thread_routine)
-    : thread_routine_(std::move(thread_routine)) {
-  RELEASE_ASSERT(Logger::Registry::initialized(), "");
-  const int rc = pthread_create(
-      &thread_handle_, nullptr,
-      [](void* arg) -> void* {
-        static_cast<ThreadImplPosix*>(arg)->thread_routine_();
-        return nullptr;
-      },
-      this);
-  RELEASE_ASSERT(rc == 0, "");
-}
+// See https://www.man7.org/linux/man-pages/man3/pthread_setname_np.3.html.
+// The maximum thread name is 16 bytes including the terminating nul byte,
+// so we need to truncate the string_view to 15 bytes.
+#define PTHREAD_MAX_LEN_INCLUDING_NULL_BYTE 16
+
+/**
+ * Wrapper for a pthread thread. We don't use std::thread because it eats exceptions and leads to
+ * unusable stack traces.
+ */
+class ThreadImplPosix : public Thread {
+public:
+  ThreadImplPosix(std::function<void()> thread_routine, OptionsOptConstRef options)
+      : thread_routine_(std::move(thread_routine)) {
+    if (options) {
+      name_ = (std::string(options->name_.substr(0, PTHREAD_MAX_LEN_INCLUDING_NULL_BYTE - 1)));
+    }
+    RELEASE_ASSERT(Logger::Registry::initialized(), "");
+    const int rc = pthread_create(
+        &thread_handle_, nullptr,
+        [](void* arg) -> void* {
+          static_cast<ThreadImplPosix*>(arg)->thread_routine_();
+          return nullptr;
+        },
+        this);
+    RELEASE_ASSERT(rc == 0, "");
+
+#ifdef __linux__
+    // If the name was not specified, get it from the OS. If the name was
+    // specified, write it into the thread, and assert that the OS sees it the
+    // same way.
+    if (name_.empty()) {
+      getNameFromOS(name_);
+    } else {
+      const int set_name_rc = pthread_setname_np(thread_handle_, name_.c_str());
+      if (set_name_rc != 0) {
+        ENVOY_LOG_MISC(trace, "Error {} setting name `{}': {}", set_name_rc, name_,
+                       strerror(set_name_rc));
+      } else {
+#ifndef NDEBUG
+        std::string check_name;
+        if (getNameFromOS(check_name)) {
+          ASSERT(check_name == name_,
```

Comment:
The extra #if seems unneeded. Can you make `getNameFromOS()` return a `std::string`, and check for equality in the `ASSERT` so that no extra #if is needed here?

Reply:
it's a little tricky because it is valid for getNameFromOS to fail. That will happen if the thread exits after the call to `pthread_setname_np` and before the call to `pthread_getname_np`. This seemed like the easiest way to assert that, if we can get the name at all, it's what we expect.

If you have other ideas of how to structure this better I'm happy to try, but this seemed like the simplest approach so far.

Reply:
Regardless of the ifdef, I'd say make it return an optional<std::string>. And then I think you can change it to `ASSERT(!getNameFromOS().valid() || getNameFromOS() == name_)`

Reply:
yup that's better; actually i had done something similar to this but I'll switch to optional.

Reply:
actually I like what I have better at both call-sites. I don't like calling getNameFromOS() twice as it can fail due to subprocess existing between the two calls.